Icecast HTTP Protocol Error On Radio Reference Service HTTP 1.0 403 Error

This may be an issue that the HTTP handler does not have HTTP 1.0 message parsing available.

2022-12-24 05:19:44.391 ERROR i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - HTTP protocol decoder error  [38MB/68MB 56%]
org.apache.mina.filter.codec.ProtocolDecoderException: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1 (Hexdump: 48 54 54 50 2F 31 2E 30 20 34 30 33 20 46 6F 72 62 69 64 64 65 6E 0D 0A 53 65 72 76 65 72 3A 20 49 63 65 63 61 73 74 20 32 2E 34 2E 34 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 43 6C 6F 73 65 0D 0A 44 61 74 65 3A 20 53 61 74 2C 20 32 34 20 44 65 63 20 32 30 32 32 20 31 30 3A 31 39 3A 34 34 20 47 4D 54 0D 0A 43 6F 6E 74 65 6E 74 2D 54 79 70 65 3A 20 74 65 78 74 2F 70 6C 61 69 6E 3B 20 63 68 61 72 73 65 74 3D 75 74 66 2D 38 0D 0A 43 61 63 68 65 2D 43 6F 6E 74 72 6F 6C 3A 20 6E 6F 2D 63 61 63 68 65 2C 20 6E 6F 2D 73 74 6F 72 65 0D 0A 45 78 70 69 72 65 73 3A 20 4D 6F 6E 2C 20 32 36 20 4A 75 6C 20 31 39 39 37 20 30 35 3A 30 30 3A 30 30 20 47 4D 54 0D 0A 50 72 61 67 6D 61 3A 20 6E 6F 2D 63 61 63 68 65 0D 0A 41 63 63 65 73 73 2D 43 6F 6E 74 72 6F 6C 2D 41 6C 6C 6F 77 2D 4F 72 69 67 69 6E 3A 20 2A 0D 0A 0D 0A 4D 6F 75 6E 74 70 6F 69 6E 74 20 69 6E 20 75 73 65)
	at org.apache.mina.filter.codec.ProtocolCodecFilter.messageReceived(ProtocolCodecFilter.java:255)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.callNextMessageReceived(DefaultIoFilterChain.java:650)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.access$1300(DefaultIoFilterChain.java:49)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain$EntryImpl$1.messageReceived(DefaultIoFilterChain.java:1128)
	at org.apache.mina.core.filterchain.IoFilterAdapter.messageReceived(IoFilterAdapter.java:122)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.callNextMessageReceived(DefaultIoFilterChain.java:650)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.fireMessageReceived(DefaultIoFilterChain.java:643)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor.read(AbstractPollingIoProcessor.java:539)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor.access$1200(AbstractPollingIoProcessor.java:68)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.process(AbstractPollingIoProcessor.java:1224)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.process(AbstractPollingIoProcessor.java:1213)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:683)
	at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at org.apache.mina.http.HttpClientDecoder.parseHttpReponseHead(HttpClientDecoder.java:242)
	at org.apache.mina.http.HttpClientDecoder.decode(HttpClientDecoder.java:114)
	at org.apache.mina.filter.codec.ProtocolCodecFilter.messageReceived(ProtocolCodecFilter.java:247)
	... 15 common frames omitted

Debug logger enabled:


```
2022-12-24 05:35:30.978 INFO  i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - CREATED  [32MB/68MB 47%]
2022-12-24 05:35:30.978 INFO  i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - OPENED  [32MB/68MB 47%]
2022-12-24 05:35:30.979 INFO  i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - SENT: HTTP REQUEST METHOD: PUT
VERSION: HTTP/1.1
PATH: /stream
QUERY:
--- HEADER --- 
Authorization:Basic c291cmNlOmRlbm55
ice-genre:My Genre
Accept:*/*
ice-name:Local
ice-audio-info:samplerate=8000;quality=7;channels=1
User-Agent:sdrtrunk
ice-public:0
ice-bitrate:16
ice-description:Denny's Local Icecast2
ice-url:http://localhost:8000
icy-metaint:2000
Content-Type:audio/mpeg
--- PARAMETERS --- 
  [32MB/68MB 47%]
2022-12-24 05:35:30.983 INFO  i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - RECEIVED: HeapBuffer@4d9c6176[pos=0 lim=288 cap=2048: 48 54 54 50 2F 31 2E 30 20 34 30 33 20 46 6F 72]  [32MB/68MB 47%]
2022-12-24 05:35:30.984 ERROR i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - HTTP protocol decoder error  [32MB/68MB 47%]
org.apache.mina.filter.codec.ProtocolDecoderException: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1 (Hexdump: 48 54 54 50 2F 31 2E 30 20 34 30 33 20 46 6F 72 62 69 64 64 65 6E 0D 0A 53 65 72 76 65 72 3A 20 49 63 65 63 61 73 74 20 32 2E 34 2E 34 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 43 6C 6F 73 65 0D 0A 44 61 74 65 3A 20 53 61 74 2C 20 32 34 20 44 65 63 20 32 30 32 32 20 31 30 3A 33 35 3A 33 30 20 47 4D 54 0D 0A 43 6F 6E 74 65 6E 74 2D 54 79 70 65 3A 20 74 65 78 74 2F 70 6C 61 69 6E 3B 20 63 68 61 72 73 65 74 3D 75 74 66 2D 38 0D 0A 43 61 63 68 65 2D 43 6F 6E 74 72 6F 6C 3A 20 6E 6F 2D 63 61 63 68 65 2C 20 6E 6F 2D 73 74 6F 72 65 0D 0A 45 78 70 69 72 65 73 3A 20 4D 6F 6E 2C 20 32 36 20 4A 75 6C 20 31 39 39 37 20 30 35 3A 30 30 3A 30 30 20 47 4D 54 0D 0A 50 72 61 67 6D 61 3A 20 6E 6F 2D 63 61 63 68 65 0D 0A 41 63 63 65 73 73 2D 43 6F 6E 74 72 6F 6C 2D 41 6C 6C 6F 77 2D 4F 72 69 67 69 6E 3A 20 2A 0D 0A 0D 0A 4D 6F 75 6E 74 70 6F 69 6E 74 20 69 6E 20 75 73 65)
	at org.apache.mina.filter.codec.ProtocolCodecFilter.messageReceived(ProtocolCodecFilter.java:255)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.callNextMessageReceived(DefaultIoFilterChain.java:650)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.access$1300(DefaultIoFilterChain.java:49)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain$EntryImpl$1.messageReceived(DefaultIoFilterChain.java:1128)
	at org.apache.mina.filter.logging.LoggingFilter.messageReceived(LoggingFilter.java:216)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.callNextMessageReceived(DefaultIoFilterChain.java:650)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.access$1300(DefaultIoFilterChain.java:49)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain$EntryImpl$1.messageReceived(DefaultIoFilterChain.java:1128)
	at org.apache.mina.core.filterchain.IoFilterAdapter.messageReceived(IoFilterAdapter.java:122)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.callNextMessageReceived(DefaultIoFilterChain.java:650)
	at org.apache.mina.core.filterchain.DefaultIoFilterChain.fireMessageReceived(DefaultIoFilterChain.java:643)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor.read(AbstractPollingIoProcessor.java:539)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor.access$1200(AbstractPollingIoProcessor.java:68)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.process(AbstractPollingIoProcessor.java:1224)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.process(AbstractPollingIoProcessor.java:1213)
	at org.apache.mina.core.polling.AbstractPollingIoProcessor$Processor.run(AbstractPollingIoProcessor.java:683)
	at org.apache.mina.util.NamePreservingRunnable.run(NamePreservingRunnable.java:64)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
	at org.apache.mina.http.HttpClientDecoder.parseHttpReponseHead(HttpClientDecoder.java:242)
	at org.apache.mina.http.HttpClientDecoder.decode(HttpClientDecoder.java:114)
	at org.apache.mina.filter.codec.ProtocolCodecFilter.messageReceived(ProtocolCodecFilter.java:247)
	... 19 common frames omitted
2022-12-24 05:35:30.985 INFO  i.g.d.a.b.AudioStreamingBroadcaster - [Local] status: Disconnected  [32MB/68MB 47%]
2022-12-24 05:35:30.986 INFO  i.g.d.a.b.i.IcecastHTTPAudioBroadcaster - CLOSED  [32MB/68MB 47%]
2022-12-24 05:35:32.836 INFO  i.g.d.a.b.AudioStreamingBroadcaster - [SDRTrunk Test Feed #1] status: Disconnected  [38MB/76MB 51%]
2022-12-24 05:35:32.866 INFO  i.g.d.a.b.AudioStreamingBroadcaster - [SDRTrunk Test Feed #2] status: Disconnected  [38MB/76MB 51%]


```

--- a/src/main/java/io/github/dsheirer/audio/broadcast/AudioStreamingBroadcaster.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/AudioStreamingBroadcaster.java
@@ -25,19 +25,15 @@ import io.github.dsheirer.audio.convert.MP3FrameTools;
 import io.github.dsheirer.audio.convert.MP3Setting;
 import io.github.dsheirer.identifier.IdentifierCollection;
 import io.github.dsheirer.util.ThreadPool;
-import org.apache.commons.math3.util.FastMath;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.nio.file.Files;
-import java.util.List;
 import java.util.Queue;
 import java.util.concurrent.LinkedTransferQueue;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public abstract class AudioStreamingBroadcaster<T extends BroadcastConfiguration> extends AbstractAudioBroadcaster<T>
 {
@@ -255,7 +251,6 @@ public abstract class AudioStreamingBroadcaster<T extends BroadcastConfiguration
     public boolean canConnect()
     {
         BroadcastState state = getBroadcastState();
-
         return state != BroadcastState.CONNECTED && !state.isErrorState();
     }
 

--- a/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastModel.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastModel.java
@@ -26,13 +26,6 @@ import io.github.dsheirer.properties.SystemProperties;
 import io.github.dsheirer.sample.Broadcaster;
 import io.github.dsheirer.sample.Listener;
 import io.github.dsheirer.util.ThreadPool;
-import javafx.collections.FXCollections;
-import javafx.collections.ObservableList;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import javax.swing.*;
-import javax.swing.table.AbstractTableModel;
 import java.io.IOException;
 import java.nio.file.DirectoryStream;
 import java.nio.file.Files;
@@ -48,6 +41,12 @@ import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.swing.table.AbstractTableModel;
 
 public class BroadcastModel extends AbstractTableModel implements Listener<AudioRecording>
 {
@@ -58,7 +57,7 @@ public class BroadcastModel extends AbstractTableModel implements Listener<Audio
 
     private static final String UNIQUE_NAME_REGEX = "(.*)\\((\\d*)\\)";
 
-    public static final int COLUMN_SERVER_ICON = 0;
+    public static final int COLUMN_BROADCAST_SERVER_TYPE = 0;
     public static final int COLUMN_STREAM_NAME = 1;
     public static final int COLUMN_BROADCASTER_STATUS = 2;
     public static final int COLUMN_BROADCASTER_QUEUE_SIZE = 3;
@@ -67,7 +66,7 @@ public class BroadcastModel extends AbstractTableModel implements Listener<Audio
     public static final int COLUMN_BROADCASTER_ERROR_COUNT = 6;
 
     public static final String[] COLUMN_NAMES = new String[]
-        {"Streaming", "Name", "Status", "Queued", "Streamed/Uploaded", "Aged Off", "Upload Error"};
+        {"Stream Type", "Name", "Status", "Queued", "Streamed/Uploaded", "Aged Off", "Upload Error"};
 
     private ObservableList<ConfiguredBroadcast> mConfiguredBroadcasts =
         FXCollections.observableArrayList(ConfiguredBroadcast.extractor());
@@ -575,15 +574,8 @@ public class BroadcastModel extends AbstractTableModel implements Listener<Audio
                 {
                     switch(columnIndex)
                     {
-                        case COLUMN_SERVER_ICON:
-                            String iconPath = configuredBroadcast.getBroadcastConfiguration()
-                                .getBroadcastServerType().getIconPath();
-
-                            if(iconPath != null && mIconModel != null)
-                            {
-                                return mIconModel.getScaledIcon(iconPath, 14);
-                            }
-                            break;
+                        case COLUMN_BROADCAST_SERVER_TYPE:
+                            return configuredBroadcast.getBroadcastServerType();
                         case COLUMN_STREAM_NAME:
                             return configuredBroadcast.getBroadcastConfiguration().getName();
                         case COLUMN_BROADCASTER_STATUS:
@@ -653,8 +645,8 @@ public class BroadcastModel extends AbstractTableModel implements Listener<Audio
             case COLUMN_BROADCASTER_STREAMED_COUNT:
             case COLUMN_BROADCASTER_ERROR_COUNT:
                 return Integer.class;
-            case COLUMN_SERVER_ICON:
-                return ImageIcon.class;
+            case COLUMN_BROADCAST_SERVER_TYPE:
+                return BroadcastServerType.class;
             case COLUMN_STREAM_NAME:
             default:
                 return String.class;

--- a/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastServerType.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastServerType.java
@@ -1,6 +1,6 @@
-/*******************************************************************************
- * sdrtrunk
- * Copyright (C) 2014-2016 Dennis Sheirer
+/*
+ * *****************************************************************************
+ * Copyright (C) 2014-2022 Dennis Sheirer
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -14,8 +14,8 @@
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>
- *
- ******************************************************************************/
+ * ****************************************************************************
+ */
 package io.github.dsheirer.audio.broadcast;
 
 public enum BroadcastServerType
@@ -23,17 +23,17 @@ public enum BroadcastServerType
     /**
      * Broadcastify feeds (ie streaming) service
      */
-    BROADCASTIFY("Broadcastify Feed", "/images/broadcastify.png"), //Icecast Server 2.3.2
+    BROADCASTIFY("Broadcastify Feed", "images/broadcastify.png"), //Icecast Server 2.3.2
 
     /**
      * Broadcastify calls - completed audio recording push service
      */
-    BROADCASTIFY_CALL("Broadcastify Call", "/images/broadcastify.png"),
+    BROADCASTIFY_CALL("Broadcastify Call", "images/broadcastify.png"),
 
-    ICECAST_HTTP("Icecast 2 (v2.4+)", "/images/icecast.png"),
-    ICECAST_TCP("Icecast (v2.3)", "/images/icecast.png"),
-    SHOUTCAST_V1("Shoutcast v1.x", "/images/shoutcast.png"),
-    SHOUTCAST_V2("Shoutcast v2.x", "/images/shoutcast.png"),
+    ICECAST_HTTP("Icecast 2 (v2.4+)", "images/icecast.png"),
+    ICECAST_TCP("Icecast (v2.3)", "images/icecast.png"),
+    SHOUTCAST_V1("Shoutcast v1.x", "images/shoutcast.png"),
+    SHOUTCAST_V2("Shoutcast v2.x", "images/shoutcast.png"),
     UNKNOWN("Unknown", null);
 
     private String mLabel;

--- a/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastState.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastState.java
@@ -1,6 +1,6 @@
-/*******************************************************************************
- * sdrtrunk
- * Copyright (C) 2014-2016 Dennis Sheirer
+/*
+ * *****************************************************************************
+ * Copyright (C) 2014-2022 Dennis Sheirer
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -14,8 +14,8 @@
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>
- *
- ******************************************************************************/
+ * ****************************************************************************
+ */
 package io.github.dsheirer.audio.broadcast;
 
 public enum BroadcastState
@@ -68,7 +68,7 @@ public enum BroadcastState
     /**
      * Specified mount point is already in use
      */
-    MOUNT_POINT_IN_USE("Mount Point In Use", false),
+    MOUNT_POINT_IN_USE("Mount Point In Use", true),
 
     /**
      * Network is unavailable or the server address cannot be resolved
@@ -103,10 +103,10 @@ public enum BroadcastState
     private String mLabel;
     private boolean mErrorState;
 
-    private BroadcastState(String label, boolean connected)
+    BroadcastState(String label, boolean errorState)
     {
         mLabel = label;
-        mErrorState = connected;
+        mErrorState = errorState;
     }
 
     public String toString()

--- a/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastStatusPanel.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/BroadcastStatusPanel.java
@@ -1,7 +1,6 @@
 /*
- * ******************************************************************************
- * sdrtrunk
- * Copyright (C) 2014-2019 Dennis Sheirer
+ * *****************************************************************************
+ * Copyright (C) 2014-2022 Dennis Sheirer
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -15,23 +14,29 @@
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>
- * *****************************************************************************
+ * ****************************************************************************
  */
 package io.github.dsheirer.audio.broadcast;
 
+import io.github.dsheirer.icon.Icon;
+import io.github.dsheirer.icon.IconModel;
 import io.github.dsheirer.preference.UserPreferences;
 import io.github.dsheirer.preference.swing.JTableColumnWidthMonitor;
+import java.awt.Color;
+import java.awt.Component;
 import net.miginfocom.swing.MigLayout;
 
+import javax.swing.ImageIcon;
 import javax.swing.JLabel;
 import javax.swing.JPanel;
 import javax.swing.JScrollPane;
 import javax.swing.JTable;
 import javax.swing.SwingConstants;
 import javax.swing.table.DefaultTableCellRenderer;
-import java.awt.Color;
-import java.awt.Component;
 
+/**
+ * Table of broadcast streams and statuses.
+ */
 public class BroadcastStatusPanel extends JPanel
 {
     private JTable mTable;
@@ -41,6 +46,12 @@ public class BroadcastStatusPanel extends JPanel
     private UserPreferences mUserPreferences;
     private String mPreferenceKey;
 
+    /**
+     * Constructs an instance
+     * @param broadcastModel to access the streams
+     * @param userPreferences for configuring the panel
+     * @param preferenceKey to store column preferences for this panel.
+     */
     public BroadcastStatusPanel(BroadcastModel broadcastModel, UserPreferences userPreferences, String preferenceKey)
     {
         mBroadcastModel = broadcastModel;
@@ -65,11 +76,43 @@ public class BroadcastStatusPanel extends JPanel
         renderer.setHorizontalAlignment(SwingConstants.CENTER);
 
         mTable.getColumnModel().getColumn(BroadcastModel.COLUMN_BROADCASTER_STATUS).setCellRenderer(new StatusCellRenderer());
+        mTable.getColumnModel().getColumn(BroadcastModel.COLUMN_BROADCAST_SERVER_TYPE).setCellRenderer(new ServerTypeRenderer());
         mColumnWidthMonitor = new JTableColumnWidthMonitor(mUserPreferences, mTable, mPreferenceKey);
 
         mScrollPane = new JScrollPane(mTable);
 
         add(mScrollPane);
+    }
+
+    public class ServerTypeRenderer extends DefaultTableCellRenderer
+    {
+        public ServerTypeRenderer()
+        {
+            setOpaque(true);
+            setHorizontalAlignment(SwingConstants.CENTER);
+        }
+
+        @Override
+        public Component getTableCellRendererComponent(JTable table, Object value, boolean isSelected, boolean hasFocus, int row, int column)
+        {
+            JLabel component = (JLabel)super.getTableCellRendererComponent(table, value, isSelected, hasFocus, row, column);
+
+            if(value instanceof BroadcastServerType broadcastServerType)
+            {
+                component.setText(broadcastServerType.toString());
+                Icon icon = new Icon("empty", broadcastServerType.getIconPath());
+                ImageIcon imageIcon = icon.getIcon();
+                ImageIcon scaledIcon = IconModel.getScaledIcon(imageIcon, 13);
+                component.setIcon(scaledIcon);
+            }
+            else
+            {
+                component.setText(null);
+                component.setIcon(null);
+            }
+
+            return component;
+        }
     }
 
     /**

--- a/src/main/java/io/github/dsheirer/audio/broadcast/icecast/IcecastHTTPAudioBroadcaster.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/icecast/IcecastHTTPAudioBroadcaster.java
@@ -20,13 +20,20 @@ package io.github.dsheirer.audio.broadcast.icecast;
 
 import io.github.dsheirer.alias.AliasModel;
 import io.github.dsheirer.audio.broadcast.BroadcastState;
-import io.github.dsheirer.audio.broadcast.icecast.IcecastMetadata;
 import io.github.dsheirer.audio.convert.InputAudioFormat;
 import io.github.dsheirer.audio.convert.MP3AudioConverter;
 import io.github.dsheirer.audio.convert.MP3Setting;
 import io.github.dsheirer.identifier.IdentifierCollection;
 import io.github.dsheirer.properties.SystemProperties;
 import io.github.dsheirer.util.ThreadPool;
+import java.net.ConnectException;
+import java.net.InetSocketAddress;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import org.apache.mina.core.RuntimeIoException;
 import org.apache.mina.core.buffer.IoBuffer;
 import org.apache.mina.core.future.ConnectFuture;
@@ -41,13 +48,6 @@ import org.apache.mina.http.api.HttpVersion;
 import org.apache.mina.transport.socket.nio.NioSocketConnector;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import java.net.ConnectException;
-import java.net.InetSocketAddress;
-import java.util.Arrays;
-import java.util.HashMap;
-import java.util.Map;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 public class IcecastHTTPAudioBroadcaster extends IcecastAudioBroadcaster
 {
@@ -143,8 +143,9 @@ public class IcecastHTTPAudioBroadcaster extends IcecastAudioBroadcaster
                 mSocketConnector = new NioSocketConnector();
                 mSocketConnector.setConnectTimeoutCheckInterval(10000);
 
-//                mSocketConnector.getFilterChain().addLast("logger",
-//                    new LoggingFilter(IcecastHTTPAudioBroadcaster.class));
+//                LoggingFilter loggingFilter = new LoggingFilter(IcecastHTTPAudioBroadcaster.class);
+//                loggingFilter.setMessageReceivedLogLevel(LogLevel.DEBUG);
+//                mSocketConnector.getFilterChain().addLast("logger", loggingFilter);
 
                 mSocketConnector.getFilterChain().addLast("codec", new HttpClientCodec());
                 mSocketConnector.setHandler(new IcecastHTTPIOHandler());
@@ -315,10 +316,49 @@ public class IcecastHTTPAudioBroadcaster extends IcecastAudioBroadcaster
                  * intercept the out of bounds exception and then inspect the message hex dump to see if it's an
                  * OK response and then play like everything is well and good.
                  */
-                if(cause instanceof ArrayIndexOutOfBoundsException &&
-                    ((ProtocolDecoderException)throwable).getHexdump().startsWith(HTTP_1_0_OK_HEX_DUMP))
+                if(cause instanceof ArrayIndexOutOfBoundsException)
                 {
-                    setBroadcastState(BroadcastState.CONNECTED);
+                    String hexDump = ((ProtocolDecoderException)throwable).getHexdump();
+
+                    if(hexDump.startsWith(HTTP_1_0_OK_HEX_DUMP))
+                    {
+                        setBroadcastState(BroadcastState.CONNECTED);
+                    }
+                    else
+                    {
+                        HttpDumpMessage message = new HttpDumpMessage(hexDump);
+
+                        if(message.hasHttpResponseCode())
+                        {
+                            switch(message.getHttpResponseCode())
+                            {
+                                case 403: //Forbidden
+                                    if(message.toString().contains("Mountpoint in use"))
+                                    {
+                                        mLog.error("Stream [" + getStreamName() + "] - unable to connect - mountpoint in use");
+                                        setBroadcastState(BroadcastState.MOUNT_POINT_IN_USE);
+                                        disconnect();
+                                    }
+                                    else
+                                    {
+                                        mLog.error("String [" + getStreamName() + "] - HTTP protocol decoder error - message:\n\n" + message);
+                                        setBroadcastState(BroadcastState.DISCONNECTED);
+                                        disconnect();
+                                    }
+                                    break;
+                                default:
+                                    mLog.error("String [" + getStreamName() + "] - HTTP protocol decoder error - message:\n\n" + message);
+                                    setBroadcastState(BroadcastState.DISCONNECTED);
+                                    disconnect();
+                            }
+                        }
+                        else
+                        {
+                            mLog.error("HTTP protocol decoder error - message:\n\n" + message);
+                            setBroadcastState(BroadcastState.DISCONNECTED);
+                            disconnect();
+                        }
+                    }
                 }
                 else
                 {
@@ -372,6 +412,99 @@ public class IcecastHTTPAudioBroadcaster extends IcecastAudioBroadcaster
                 mLog.error("Icecast HTTP broadcaster - unrecognized message [ " + object.getClass() +
                     "] received:" + object.toString());
             }
+        }
+    }
+
+    /**
+     * Class for parsing HTTP response messages from Icecast 2.4.2+
+     */
+    public class HttpDumpMessage
+    {
+        private String mHexDump;
+        private String mMessage;
+        private int mHttpResponseCode = -1;
+
+        public HttpDumpMessage(String hexDump)
+        {
+            mHexDump = hexDump;
+
+            String[] split = mHexDump.split(" ");
+            byte[] bytes = new byte[split.length];
+
+            int pointer = 0;
+            for(String a : split)
+            {
+                try
+                {
+                    int value = Integer.parseInt(a, 16);
+                    bytes[pointer++] = (byte) (value & 0xFF);
+                }
+                catch(Exception e)
+                {
+                    pointer++;
+                }
+            }
+
+            mMessage = new String(bytes);
+
+            Pattern pattern = Pattern.compile("HTTP/1.0 (\\d{3})");
+            Matcher m = pattern.matcher(mMessage);
+
+            if(m.find())
+            {
+                try
+                {
+                    mHttpResponseCode = Integer.parseInt(m.group(1));
+                }
+                catch(Exception e)
+                {
+                    mLog.error("Unable to parse HTTP response code that was matched from message: " + m.group(1));
+                }
+            }
+        }
+
+        /**
+         * Indicates if an HTTP response code was parsed from the message
+         * @return true if it was parsed.
+         */
+        public boolean hasHttpResponseCode()
+        {
+            return mHttpResponseCode != -1;
+        }
+
+        /**
+         * Response code parsed from the hexdump message
+         * @return value.
+         */
+        public int getHttpResponseCode()
+        {
+            return mHttpResponseCode;
+        }
+
+        @Override
+        public String toString()
+        {
+            return mMessage;
+        }
+    }
+
+    public static void main(String[] args)
+    {
+        String text = "HTTP/1.0 403 Forbidden";
+
+        Pattern pattern = Pattern.compile("HTTP/1.0 (\\d{3})");
+
+        Matcher m = pattern.matcher(text);
+
+        if(m.find())
+        {
+            mLog.info("Matching Group Count: " + m.groupCount());
+            mLog.info("Match 0: " + m.group(0));
+            mLog.info("Match 1: " + m.group(1));
+        }
+        else
+        {
+            mLog.info("No Match");
         }
     }
 }

--- a/src/main/java/io/github/dsheirer/audio/broadcast/icecast/IcecastTCPAudioBroadcaster.java
+++ b/src/main/java/io/github/dsheirer/audio/broadcast/icecast/IcecastTCPAudioBroadcaster.java
@@ -21,13 +21,19 @@ package io.github.dsheirer.audio.broadcast.icecast;
 import io.github.dsheirer.alias.AliasModel;
 import io.github.dsheirer.audio.broadcast.BroadcastState;
 import io.github.dsheirer.audio.broadcast.icecast.codec.IcecastCodecFactory;
-import io.github.dsheirer.audio.broadcast.icecast.IcecastMetadata;
 import io.github.dsheirer.audio.convert.InputAudioFormat;
 import io.github.dsheirer.audio.convert.MP3AudioConverter;
 import io.github.dsheirer.audio.convert.MP3Setting;
 import io.github.dsheirer.identifier.IdentifierCollection;
 import io.github.dsheirer.properties.SystemProperties;
 import io.github.dsheirer.util.ThreadPool;
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.net.SocketException;
+import java.nio.channels.UnresolvedAddressException;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import org.apache.mina.core.RuntimeIoException;
 import org.apache.mina.core.future.ConnectFuture;
 import org.apache.mina.core.service.IoHandlerAdapter;
@@ -36,14 +42,6 @@ import org.apache.mina.filter.codec.ProtocolCodecFilter;
 import org.apache.mina.transport.socket.nio.NioSocketConnector;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import java.io.IOException;
-import java.net.InetSocketAddress;
-import java.net.SocketException;
-import java.nio.channels.UnresolvedAddressException;
-import java.util.Arrays;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 public class IcecastTCPAudioBroadcaster extends IcecastAudioBroadcaster
 {
@@ -215,7 +213,13 @@ public class IcecastTCPAudioBroadcaster extends IcecastAudioBroadcaster
         }
         else
         {
-            setBroadcastState(BroadcastState.DISCONNECTED);
+            //Only set broadcast state to disconnected if it's not already in an error state, to prevent a restart.  We
+            //want to preserve the error state that got us here, so the user can see it.
+            if(!getBroadcastState().isErrorState())
+            {
+                setBroadcastState(BroadcastState.DISCONNECTED);
+            }
+
             mLastConnectionAttempt = System.currentTimeMillis();
         }
     }
@@ -317,6 +321,7 @@ public class IcecastTCPAudioBroadcaster extends IcecastAudioBroadcaster
             {
                 String message = (String) object;
 
+                mLog.info("Stream [" + getStreamName() + "] message received: " + message);
                 if(message != null && !message.trim().isEmpty())
                 {
                     if(message.startsWith("HTTP/1.0 200 OK"))

--- a/src/main/java/io/github/dsheirer/icon/IconModel.java
+++ b/src/main/java/io/github/dsheirer/icon/IconModel.java
@@ -1,18 +1,20 @@
 /*
- * *********************************************************************************************************************
- * sdrtrunk
- * Copyright (C) 2014-2017 Dennis Sheirer
+ * *****************************************************************************
+ * Copyright (C) 2014-2022 Dennis Sheirer
  *
- * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public
- * License as published by  the Free Software Foundation, either version 3 of the License, or  (at your option) any
- * later version.
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
  *
- * This program is distributed in the hope that it will be useful,  but WITHOUT ANY WARRANTY; without even the implied
- * warranty of  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
  *
- * You should have received a copy of the GNU General Public License  along with this program.
- * If not, see <http://www.gnu.org/licenses/>
- * *********************************************************************************************************************
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ * ****************************************************************************
  */
 package io.github.dsheirer.icon;
 
@@ -22,15 +24,6 @@ import com.fasterxml.jackson.dataformat.xml.JacksonXmlModule;
 import com.fasterxml.jackson.dataformat.xml.XmlMapper;
 import io.github.dsheirer.properties.SystemProperties;
 import io.github.dsheirer.util.ThreadPool;
-import javafx.beans.property.SimpleStringProperty;
-import javafx.beans.property.StringProperty;
-import javafx.collections.FXCollections;
-import javafx.collections.ListChangeListener;
-import javafx.collections.ObservableList;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
-
-import javax.swing.ImageIcon;
 import java.awt.Image;
 import java.io.IOException;
 import java.io.InputStream;
@@ -43,6 +36,15 @@ import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import javafx.beans.property.SimpleStringProperty;
+import javafx.beans.property.StringProperty;
+import javafx.collections.FXCollections;
+import javafx.collections.ListChangeListener;
+import javafx.collections.ObservableList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.swing.ImageIcon;
 
 public class IconModel
 {
@@ -241,7 +243,6 @@ public class IconModel
         if(path != null)
         {
             Icon icon = new Icon("", path);
-
             return getScaledIcon(icon.getIcon(), height);
         }
 

